rcl/include/rcl/remap.h

```diff
@@ -0,0 +1,209 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__REMAP_H_
+#define RCL__REMAP_H_
+
+#include "rcl/allocator.h"
+#include "rcl/arguments.h"
+#include "rcl/macros.h"
+#include "rcl/types.h"
+#include "rcl/visibility_control.h"
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+/// Remaps a topic name based on given rules
+/**
+ * The supplied topic name must have already been expanded to a fully qualified name.
+ * \sa rcl_expand_topic_name()
+ *
+ * If the node has been given arguments then the remap rules from those will be checked first.
+ * If no rules matched, then global remap rules will be checked if the node has not also been
+ * instructed to ignore global arguments.
+ *
+ * Remap rules are checked in the order they were given.
+ * Processing stops when a remap rule has been matched or there are no more rules.
```

Comment:
> if not, should there be?

By multiple passes do you mean using every rule in order using the result of the previous rule rather than stopping when one matches?

An argument against is that remapping stopped at the first rule in ROS 1, so a set of rules that work in ROS 1 may not work in ROS 2. Another argument against is that it adds an extra rule to swap topic names: `foo:=temporary_name bar:=foo temporary_name:=bar`.

An argument for is that dynamic remapping could take advantage of it if publishers/clients/etc store the original name. Dynamic remapping would add a new rule to the end of the list and then have all rules re-applied. Names that come up after a dynamic rule is added could be remapped as if the dynamic rule was a static one all along.

Reply:
I added the corner cases you mentioned and expanded the documentation in c04c2ebc7d21ec0a1d7fc99ff7d51320bcf22f07

Reply:
> By multiple passes do you mean using every rule in order using the result of the previous rule rather than stopping when one matches?

Yeah, that's what I meant, but I was thinking of a different way to describe it. Basically, you could continue applying rules, in order, to a single input until it matches none, but then you need to detect cycles...

I don't think we should support it, I just think it would be nice to point out that we don't do that explicitly. It's one of those things that someone could easily assume is happening without realizing that it doesn't because it's complicated and we didn't implement it.

I think the documentation changes you made are sufficient, thanks!